blockchain/indexers/common.go

```diff
@@ -173,3 +171,788 @@ func interruptRequested(ctx context.Context) bool {
+		if !ok {
+			return fmt.Errorf("no subscription found with id %s", s.prerequisite)
+		}
+
+		prereq.mtx.Lock()
+		prereq.dependency = nil
+		prereq.mtx.Unlock()
+
+		return nil
+	}
+
+	// If the subscription has a dependency, stop it as well.
+	if s.dependency != nil {
+		err := s.dependency.Stop()
+		if err != nil {
+			return err
+		}
+	}
+
+	// If the subscription is independent, remove it from the
+	// index subscriber's subscriptions.
+	delete(s.subscriber.subscriptions, s.id)
+
+	return nil
+}
+
+// publishIndexNtfn sends the provided index notification on the channel
+// associated with the subscription.
+func (s *IndexSubscription) publishIndexNtfn(ntfn *IndexNtfn) {
+	select {
+	case s.c <- ntfn:
+	default:
+		// Non-blocking send fallthrough.
+	}
+}
+
+// IndexSubscriber subscribes clients for index updates.
+type IndexSubscriber struct {
+	c             chan *IndexNtfn
+	subscriptions map[string]*IndexSubscription
+	mtx           sync.Mutex
+	ctx           context.Context
+	cancel        context.CancelFunc
+}
+
+// NewIndexSubsciber creates a new index subscriber. It also starts
+// handler for incoming index update subscriptions.
+func NewIndexSubscriber(sCtx context.Context) *IndexSubscriber {
+	ctx, cancel := context.WithCancel(context.Background())
+	s := &IndexSubscriber{
+		c:             make(chan *IndexNtfn, bufferSize),
+		subscriptions: make(map[string]*IndexSubscription),
+		ctx:           ctx,
+		cancel:        cancel,
+	}
+
+	go s.handleNotifications(sCtx)
+
+	return s
+}
+
+// Subscribe subscribes an index for updates.  The returned index subscription
+// has functions to retrieve a channel that produces a stream of index updates
+// and to stop the stream when the caller no longer wishes to receive updates.
+func (s *IndexSubscriber) Subscribe(index Indexer, prerequisite string) (*IndexSubscription, error) {
+	sub := newIndexSubscription(s, index, prerequisite)
+
+	// If the subscription has a prequisite, find it and set the subscription
+	// as a dependency.
+	if prerequisite != noPrereqs {
+		s.mtx.Lock()
+		prereq, ok := s.subscriptions[prerequisite]
+		s.mtx.Unlock()
+		if !ok {
+			return nil, fmt.Errorf("no subscription found with id %s", prerequisite)
+		}
+
+		prereq.mtx.Lock()
+		defer prereq.mtx.Unlock()
+
+		if prereq.dependency != nil {
+			return nil, fmt.Errorf("%s already has a dependency set: %s",
+				prereq.id, prereq.dependency.id)
+		}
+
+		prereq.dependency = sub
+
+		return sub, nil
+	}
+
+	// If the subscription does not have a prerequisite, add it to the index
+	// subscriber's subscriptions.
+	s.mtx.Lock()
+	s.subscriptions[sub.id] = sub
+	s.mtx.Unlock()
+
+	return sub, nil
+}
+
+// Notify relays an index notification to subscribed indexes for processing.
+func (s *IndexSubscriber) Notify(ntfn *IndexNtfn) {
+	s.mtx.Lock()
+	clients := len(s.subscriptions)
+	s.mtx.Unlock()
+
+	// Only relay notifications when there are subscribed indexes
+	// to be notified.
+	if clients > 0 {
+		s.c <- ntfn
```

Comment:
I see what you're referring to however the `IndexSubscriber` would need a goroutine to coordinate the shutdown process for the indexes instead of passing the server context individually to each index. As a result relaying index notifications is also being done there. Relaying index notifications from `Notify` could the block the block connection/disconnection process for every index being notified as well.

Reply:
OK, no major issue, just thought that it's maybe a little harder to conceptually follow.

It is still going to ultimately block on `Notify`, it's just that you have two layers of buffers.  For example:

- If an `IndexSubscription` buffer is full, then the `IndexSubscriber` is going to block in `publishIndexNtfn` (where it writes to that `IndexSubscription` channel).
- If `IndexSubscriber` is blocked, it's not going to be dequeuing from its own buffer, which will then fill up.
- When the `IndexSubscriber` buffer is full, `Notify` will block when trying to write to the channel.

This isn't an issue since it has to block at some point if the index gets far enough behind, but just clarifying that point,